File: Pico_Wheel/Test Codes/Buttons.py
#include <stdio.h>
#include "pico/stdlib.h"
#include "hardware/spi.h"
#include "hardware/gpio.h"

// Define SPI and button configuration
#define SPI_PORT spi0
#define PIN_MISO 16
#define PIN_CS 17
#define PIN_SCK 18
#define PIN_MOSI 19
#define BUTTON_PIN 2

// Struct to hold button press state
typedef struct {
    uint32_t timestamp;
    bool button_pressed;
} ButtonState;

int main() {
    // Initialize GPIO and SPI
    stdio_init_all();
    gpio_init(BUTTON_PIN);
    gpio_set_dir(BUTTON_PIN, GPIO_IN);
    gpio_pull_up(BUTTON_PIN); // Pull-up resistor for button

    spi_init(SPI_PORT, 1000000); // Initialize SPI at 1 MHz
    gpio_set_function(PIN_MISO, GPIO_FUNC_SPI);
    gpio_set_function(PIN_CS, GPIO_FUNC_SPI);
    gpio_set_function(PIN_SCK, GPIO_FUNC_SPI);
    gpio_set_function(PIN_MOSI, GPIO_FUNC_SPI);

    // Enable SPI slave mode
    gpio_set_dir(PIN_CS, GPIO_IN);

    // Variables for button state and timing
    ButtonState button_state;
    button_state.timestamp = 0;
    button_state.button_pressed = false;

    absolute_time_t last_check_time = get_absolute_time();
    const uint32_t check_interval_us = 1000; // 1 ms

    uint8_t buffer[sizeof(ButtonState)];

    while (true) {
        // Check button state every 1 ms
        if (absolute_time_diff_us(last_check_time, get_absolute_time()) >= check_interval_us) {
            last_check_time = get_absolute_time();
            button_state.timestamp++;
            button_state.button_pressed = !gpio_get(BUTTON_PIN); // Button pressed = LOW
        }

        // Wait for master request
        if (!gpio_get(PIN_CS)) { // Active low CS
            // Send button state as a struct
            memcpy(buffer, &button_state, sizeof(ButtonState));
            spi_write_blocking(SPI_PORT, buffer, sizeof(ButtonState));
        }
    }
    return 0;
}

File: Pico_Wheel/Test Codes/ButtonS/ButtonS.c
#include <stdio.h>
#include <string.h>
#include "pico/stdlib.h"
#include "hardware/gpio.h"

// Define button GPIO pins
#define BUTTON_PIN_1 14
#define BUTTON_PIN_2 15
#define BUTTON_PIN_3 16
#define BUTTON_PIN_4 17
#define BUTTON_PIN_5 18
#define BUTTON_PIN_6 19
#define BUTTON_PIN_7 20
#define BUTTON_PIN_8 21

// Struct to hold button press state
typedef struct {
    uint32_t timestamp;
    bool button_1_pressed;
    bool button_2_pressed;
    bool button_3_pressed;
    bool button_4_pressed;
    bool button_5_pressed;
    bool button_6_pressed;
    bool button_7_pressed;
    bool button_8_pressed;
} ButtonState;

int main() {
    // Initialize GPIO and USB Serial
    stdio_init_all();
    gpio_init(BUTTON_PIN_1);
    gpio_init(BUTTON_PIN_2);
    gpio_init(BUTTON_PIN_3);
    gpio_init(BUTTON_PIN_4);
    gpio_init(BUTTON_PIN_5);
    gpio_init(BUTTON_PIN_6);
    gpio_init(BUTTON_PIN_7);
    gpio_init(BUTTON_PIN_8);
    gpio_set_dir(BUTTON_PIN_1, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_2, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_3, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_4, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_5, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_6, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_7, GPIO_IN);
    gpio_set_dir(BUTTON_PIN_8, GPIO_IN);
    gpio_pull_up(BUTTON_PIN_1); // Pull-up resistor for button 1
    gpio_pull_up(BUTTON_PIN_2); 
    gpio_pull_up(BUTTON_PIN_3); 
    gpio_pull_up(BUTTON_PIN_4); 
    gpio_pull_up(BUTTON_PIN_5); 
    gpio_pull_up(BUTTON_PIN_6); 
    gpio_pull_up(BUTTON_PIN_7); 
    gpio_pull_up(BUTTON_PIN_8); 

    // Variables for button state and timing
    ButtonState button_state;
    button_state.timestamp = 0;
    button_state.button_1_pressed = false;
    button_state.button_2_pressed = false;
    button_state.button_3_pressed = false;
    button_state.button_4_pressed = false;
    button_state.button_5_pressed = false;
    button_state.button_6_pressed = false;
    button_state.button_7_pressed = false;
    button_state.button_8_pressed = false;

        absolute_time_t last_check_time = get_absolute_time();
        const uint32_t check_interval_us = 1000; // 1 ms

    while (true) {

        
        // Check button state every 1 ms
        if (absolute_time_diff_us(last_check_time, get_absolute_time()) >= check_interval_us) {
            last_check_time = get_absolute_time();
            button_state.timestamp++;
            button_state.button_1_pressed = !gpio_get(BUTTON_PIN_1); // Button 1 pressed = LOW
            button_state.button_2_pressed = !gpio_get(BUTTON_PIN_2); // Button 2 pressed = LOW
            button_state.button_3_pressed = !gpio_get(BUTTON_PIN_3); // Button 1 pressed = LOW
            button_state.button_4_pressed = !gpio_get(BUTTON_PIN_4); // Button 2 pressed = LOW
            button_state.button_5_pressed = !gpio_get(BUTTON_PIN_5); // Button 1 pressed = LOW
            button_state.button_6_pressed = !gpio_get(BUTTON_PIN_6); // Button 2 pressed = LOW
            button_state.button_7_pressed = !gpio_get(BUTTON_PIN_7); // Button 1 pressed = LOW
            button_state.button_8_pressed = !gpio_get(BUTTON_PIN_8); // Button 2 pressed = LOW

            // Print button state to USB terminal
            printf("Timestamp: %u, Button 1: %s, Button 2: %s, Button 3: %s, Button 4: %s, Button 5: %s, Button 6: %s, Button 7: %s, Button 8: %s\n",
                   button_state.timestamp,
                   button_state.button_1_pressed ? "Pressed" : "Released",
                   button_state.button_2_pressed ? "Pressed" : "Released",
                   button_state.button_3_pressed ? "Pressed" : "Released",
                   button_state.button_4_pressed ? "Pressed" : "Released",
                   button_state.button_5_pressed ? "Pressed" : "Released",
                   button_state.button_6_pressed ? "Pressed" : "Released",
                   button_state.button_7_pressed ? "Pressed" : "Released",
                   button_state.button_8_pressed ? "Pressed" : "Released");
        }
    }
    return 0;
}
